core/src/main/java/org/commonjava/indy/core/change/StoreContentListener.java

```diff
@@ -328,198 +192,109 @@ private int clearPath( String path, ArtifactStore origin, Set<Group> affectedGro
+            Set<Group> affected = groups;
+            if ( affected == null )
             {
-                Transfer next = toProcess.remove( 0 );
                 try
                 {
-                    Stream.of(next.list()).forEach( filename->{
-                        Transfer t = next.getChild( filename );
-                        if ( t.isDirectory() )
-                        {
-                            logger.debug( "Adding directory path for processing: {}", t.getPath() );
-                            toProcess.add( t );
-                        }
-                        else
-                        {
-                            logger.trace( "Testing file path: {}", t.getPath() );
-                            if( pathFilter.test( t.getPath() ) )
-                            {
-                                logger.trace( "Adding file path to results: {}", t.getPath() );
-                                pathAction.accept( t.getPath() );
-                            }
-                            else
-                            {
-                                logger.trace( "Skipping file path: {}", t.getPath() );
-                            }
-                        }
-                    } );
+                    affected = ( storeDataManager.query().packageType( key.getPackageType() ).getGroupsAffectedBy( key ) );
                 }
-                catch ( IOException e )
+                catch ( IndyDataException e )
                 {
-                    logger.error( String.format( "Failed to list contents of: %s. Reason: %s", next, e ), e );
+                    logger.error( "Failed to retrieve groups affected by: " + key, e );
+                    return;
                 }
             }
-        }
-        catch ( IndyWorkflowException  e )
-        {
-            logger.error( String.format( "Failed to retrieve root directory reference for: %s. Reason: %s", key, e ), e );
-        }
-        finally
-        {
-            if ( root != null )
-            {
-                root.unlock();
-            }
-        }
-    }
 
-    /**
-     * List the mergable paths in target store and clean up the paths in affected groups.
-     * @param event
-     * @param pathFilter
-     * @param deleteOriginPath
-     */
-    private void processAllPaths( final IndyStoreEvent event, Predicate<? super String> pathFilter, boolean deleteOriginPath )
-    {
-        DrainingExecutorCompletionService<Integer> clearService =
-                new DrainingExecutorCompletionService<>( cleanupExecutor );
-
-        Set<StoreKey> keys = event.getStores().stream().map( store -> store.getKey() ).collect( Collectors.toSet() );
-        keys.forEach(key->{
-            final Set<Group> groups = new HashSet<>();
+            logger.debug( "Submit clean job for origin: {}", origin );
+            final Set<Group> affectedGroups = affected;
+            Future<Integer> job = cleanupExecutor.submit(
```

Comment:
What are we doing with the job? Is that just the debugging stuff below, which is commented out?

My concern with this approach is that we might return from the user request without having completed the cleanup, which could lead to subsequent requests against the same content that is in the process of being cleared. For example, a subsequent build process might choose the wrong version of a dependency for alignment because obsolete maven-metadata.xml files haven't been deleted yet.

I wonder if we need to have some sort of status monitor that we can query to tell when the cleanup process is complete? Then, the build orchestrator could simply watch the cleanup status ID until it cleared before starting the next build in the chain...

Reply:
It seems the "job" is confusing. That is only for the debugging stuff below. (It turns out to be not needed). I will remove it.

When I made the change, I mainly consider the update from a user request on UI. In this case, async is good if it takes more than, say 30 seconds. Once the user got the response, the group definition has been updated (cleaning may be not). If he or she mistakenly issue another update-request, the diff will prevent it from being executed because the diff will tell no-change. 

When it comes to PNC build, e.g., one build adds a hosted repo to some group, and the clean-up begins. Then the second build wants to use metadata which is being cleaned ATM. I would say we don't know what to expect yet. Even we use the draining service, this does not help because the second build won't wait for the former to finish - in fact, they don't know each other. 

Adding a status API for PNC to check (before promoting a new hosted repo) might make things more complicated. Basically, they can check the status at one time, but can not guarantee that light keeps green, because other builds might join the party at any time. Essentially, they will have to lock the update at the right place and make the updating in serial. I guess this is not easy. PNC locking the whole promotion seems anti-performance. We have to control it inside Indy.

If the second build wants to update the same group, we can add a check to see if the previous clean-up has been finished. I think we can use a simple Map to hold the clean-up status and check it when we start handling an update-event. The tricky thing is that the event is handled at more than one places and future handlers may forget to check the status. Anyway, this is doable since we are the master of our code :). I will push a way to do it.

Reply:
After a second thought, I guess concurrent clean-ups won't harm us. They may just delete some same files and ISPN entries. Our code report errors but not fail the clean up itself. 
I pushed a commit to fix the allPath(). 
I suggest we merge this and keep an eye on the "build using obsolete file" issue.